src/test/beans.t.sol

```diff
@@ -0,0 +1,253 @@
+        fee = 1000000003593629043335673583
+        */
+        uint fee = uint(1000000003593629043335673583); // 12 % per year
+        beans.file(fee, fee);
+        uint loan = 1;
+        uint principal = 66 ether;
+        beans.drip(fee);
+        beans.incLoanDebt(loan, fee, principal);
+
+        checkDebt(loan, fee, 66 ether);
+
+        // on year later
+        hevm.warp(now + 365 days);
+        beans.drip(fee);
+        checkDebt(loan, fee, 73.92 ether);// 66 ether * 1,12
+    }  
+
+    function testDrip() public {
+        uint fee = uint(1000000564701133626865910626); // 5 % / day
+        beans.file(fee, fee);
+        (uint debt1, uint chi1, uint speed1, uint rho1 ) = beans.fees(fee);
+        assertEq(speed1, fee);
+        assertEq(rho1, now);
+        assertEq(debt1, 0);
+
+        // on day later
+        hevm.warp(now + 1 days);
+
+        (debt1,  chi1,  speed1,  rho1 ) = beans.fees(fee);
+        assertEq(speed1, fee);
+        assertEq(debt1, 0);
+        assertTrue(rho1 != now);
+
+        beans.drip(fee);
+
+        (uint debt2, uint chi2, uint speed2, uint rho2 ) = beans.fees(fee);
+        assertEq(speed2, fee);
+        assertEq(rho2, now);
+        assertEq(debt2, 0);
+        assertTrue(chi1 != chi2);
+    }  
+
+    function testMaxChi() public {
+        // chi is uint, max value = (2^256)-1 = 1.1579209e+77
+        // chi initial 10^27
+        uint fee = uint(1000000564701133626865910626); // 5 % / daily
+        beans.file(fee, fee);
+        hevm.warp(now + 1050 days); // 1,05 ^1050 = 1.7732257e+22
+
+        // init chi 10^27 *  1.7732257 * 10^22  ~ chi 10^49
+        // rdiv operation needs to mul chi with ONE (10^27)
+        // therefore: 10^49 * 10^27 = 10^76 < 1.1579209e+77
+        beans.drip(fee);
+    }
+    
+    function testFailChiTooHigh() public {
+        // chi is uint, max value = (2^256)-1 = 1.1579209e+77
+        // chi initial 10^27
+        uint fee = uint(1000000564701133626865910626); // 5 % / daily
+        beans.file(fee, fee);
+        hevm.warp(now + 1100 days); // 1,05 ^1100 = 2.0334288e+23
+
+        // init chi 10^27 *  2.0334288 * 10^23  ~ chi 10^50
+        // rdiv operation needs to mul chi with ONE (10^27)
+        // therefore: 10^50 * 10^27 = 10^77 same power as max value 1.1579209e+77
+        beans.drip(fee);
+    }
+
+    function testMaxDebt() public {
+        uint fee = uint(1000000564701133626865910626); // 5 % day
+        beans.file(fee, fee);
+        uint loan = 1;
+        uint principal = 1000000000  ether; // one billion 10^9 * 10^18 = 10^28
+        beans.drip(fee);
+        beans.incLoanDebt(loan, fee, principal);
+
+        // 150 days later
+        hevm.warp(now + 1050 days); // produces max ~ chi 10^49
+        // debt ~ 10^27 * 10^49 =  10^76 (max uint is 10^77)
+        beans.drip(fee);
+    }
+
+    function checkDebt(uint loan, uint fee, uint should) public {
+        uint debt = beans.debtOf(loan, fee);
```

Comment:
one edge case we could test for is updating of fee in the middle of a loan term. maybe we want to consider disallowing it?

Reply:
It is actually a feature. 

`fee`could for example be a risk group like `low` risk and `high` risk and the interest rates are changing.